src/compute/src/sink/materialized_view_v2.rs

```diff
@@ -33,5 +155,874 @@ pub(super) fn persist_sink<S>(
 where
     S: Scope<Timestamp = Timestamp>,
 {
-    todo!()
+    let mut scope = ok_collection.scope();
+    let desired = OkErr::new(ok_collection.inner, err_collection.inner);
+
+    // Read back the persist shard.
+    let (mut persist, persist_token) = persist_source(
+        &mut scope,
+        sink_id,
+        target.clone(),
+        compute_state,
+        start_signal,
+    );
+
+    // Report sink frontier updates to the `ComputeState`.
+    let sink_frontier = Rc::new(RefCell::new(Antichain::from_elem(Timestamp::MIN)));
+    let collection = compute_state.expect_collection_mut(sink_id);
+    collection.sink_write_frontier = Some(Rc::clone(&sink_frontier));
+
+    persist.ok = persist.ok.inspect_container(move |event| {
+        if let Err(frontier) = event {
+            let mut borrow = sink_frontier.borrow_mut();
+            borrow.clear();
+            borrow.extend(frontier.iter().copied());
+        }
+    });
+
+    // Determine the active worker for single-worker operators.
+    let active_worker_id = usize::cast_from(sink_id.hashed()) % scope.peers();
+
+    let operator_name = |name| format!("persist_sink({sink_id})::{name}");
+    let persist_api = |purpose| PersistApi {
+        persist_clients: Arc::clone(&compute_state.persist_clients),
+        collection: target.clone(),
+        shard_name: sink_id.to_string(),
+        purpose,
+    };
+
+    let name = operator_name("mint");
+    let (desired, descs, mint_token) = mint::render(
+        name.clone(),
+        persist_api(name),
+        as_of,
+        active_worker_id,
+        &desired,
+    );
+
+    let name = operator_name("write");
+    let (batches, write_token) =
+        write::render(name.clone(), persist_api(name), &desired, &persist, &descs);
+
+    let name = operator_name("append");
+    let append_token = append::render(
+        name.clone(),
+        persist_api(name),
+        active_worker_id,
+        &descs,
+        &batches,
+    );
+
+    Rc::new((persist_token, mint_token, write_token, append_token))
+}
+
+/// Generic wrapper around ok/err pairs (e.g. streams, frontiers), to simplify code dealing with
+/// such pairs.
+struct OkErr<O, E> {
+    ok: O,
+    err: E,
+}
+
+impl<O, E> OkErr<O, E> {
+    fn new(ok: O, err: E) -> Self {
+        Self { ok, err }
+    }
+}
+
+impl OkErr<Antichain<Timestamp>, Antichain<Timestamp>> {
+    fn new_frontiers() -> Self {
+        Self {
+            ok: Antichain::from_elem(Timestamp::MIN),
+            err: Antichain::from_elem(Timestamp::MIN),
+        }
+    }
+
+    fn frontier(&self) -> &Antichain<Timestamp> {
+        if PartialOrder::less_equal(&self.ok, &self.err) {
+            &self.ok
+        } else {
+            &self.err
+        }
+    }
+}
```

Comment:
The implementation seems odd, specifically without documentation.

Reply:
This just computes the global frontier by producing the meet of the two frontiers. We could use `Lattice::meet`, but that does more work as it isn't specialized to total times, and has to allocate a new `Antichain`.

Can you say what specifically you find odd?